`General`: Allow CSV user import by email

daniels98it opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Currently either the users' **login** or **registration number** has to be provided within the CSV file in order to import users to a course's group. It may happen that instructors possess only a list of the students' emails and therefore cannot import them using the CSV import functionality.

### Description
<!-- Describe your changes in detail -->
As  the **email** represent another attribute which can be used to uniquely identify a user, this can be integrated in the already implemented CSV import functionality, extending its use cases.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor
- 1 Course

1. Log in to Artemis
2. Navigate to Course Administration
3. Add users to one or more of the course's groups (Student, Tutors, Editors, Instructors) using the 
![image](https://user-images.githubusercontent.com/45761921/194307428-02ae6693-7e37-46e9-baa0-de6def618b4f.png)
CSV import
4. The CSV used should contain at least one row containing a user's email only

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
| Class/File | Branch | Line |
|------------|-------:|-----:|
| CourseService.java | 78% | 92% |
| ExamRegistrationService.java | 85% | 97% |
| UserService.java | 78% | 90% |

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

JohannesStoehr commented on src/test/javascript/spec/component/exam/manage/student-exams/user-import-dialog.component.spec.ts:
```diff
@@ -138,17 +138,38 @@ describe('UsersImportButtonComponent', () => {
+        const event = { target: { files: [csv] } };
+        component.onCSVFileSelect(event);
+
+        expect(component.usersToImport).toHaveLength(5);
```
Is this the intended place?

daniels98it replied:
The tests should now have been fixed 👍

JohannesStoehr reviewed (commented):
Code looks good, but there are some issues with the tests in exam-management.service.spec.ts and user-import-dialog.component.spec.ts

b-fein commented on src/test/javascript/spec/component/exam/manage/student-exams/user-import-dialog.component.spec.ts:
```diff
@@ -138,17 +138,38 @@ describe('UsersImportButtonComponent', () => {
         });
     });
 
+    describe('should read students from csv with email only', () => {
```
```suggestion
    it('should read students from csv with email only', () => {
```

daniels98it replied:
Thanks 👍

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -411,7 +411,8 @@ public List<StudentDTO> registerUsersForCourseGroup(Long courseId, List<StudentD
             var registrationNumber = studentDto.getRegistrationNumber();
             var login = studentDto.getLogin();
-            Optional<User> optionalStudent = userService.findUserAndAddToCourse(registrationNumber, courseGroupName, courseGroupRole, login);
+            var email = studentDto.getEmail();
```
💡The above Doc for this method seems to be outdated now:
```
     * Note: registration based on other user attributes (e.g. email, name, login) is currently NOT supported
```
Maybe you could adapt it?

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/user/UserService.java:
```diff
@@ -662,7 +663,16 @@ public Optional<User> findUserAndAddToCourse(String registrationNumber, String c
             }
 
-            log.warn("User with registration number '{}' and login '{}' not found in Artemis user database nor found in (TUM) LDAP", registrationNumber, login);
+            // 4) if we still cannot find the user, we try again using the email
```
💡Idea: You could greatly reduce the code duplication in this method, by writing:
```
            var optionalStudent = 
                userRepository.findUserWithGroupsAndAuthoritiesByRegistrationNumber(registrationNumber)
                    .or(() -> this.createUserFromLdap(registrationNumber))
                    .or(() -> userRepository.findUserWithGroupsAndAuthoritiesByLogin(login))
                    .or(() ->  userRepository.findUserWithGroupsAndAuthoritiesByEmail(email));
            
            
            if (optionalStudent.isPresent()) {
                var student = optionalStudent.get();
                // the newly created user needs to get the rights to access the course
                this.addUserToGroup(student, courseGroupName, courseGroupRole);
                return optionalStudent;
            }
```

daniels98it replied:
Very good refactoring suggestion, thanks 😁
Added it with https://github.com/ls1intum/Artemis/pull/5722/commits/5e90ac4b0499884f48d7b066899ab045e2aa2760

stefanwaldhauser reviewed (approved):
Great job! Just some nice suggestions :)

PhilipWerz reviewed (approved):
Everything works as it should. Tested on TS1.

Note: 
You have to at least add two new users in the csv in order for it to work

timo07k reviewed (approved):
Import works (tested for importing new students with their email). Tested on TS1.

maximiliansoelch reviewed (approved):
Tested on TS1, import works as expected 👍

stefanwaldhauser reviewed (dismissed):
Reapprove

daniels98it commented:
> Currently, only works for > 1 email entries. When only 1 email is in CSV file, the following error is displayed: ![image](https://user-images.githubusercontent.com/10652693/203321701-92dccd25-b1af-4733-9fc5-c26e12c4642d.png)

@evgeny1984 could you post the content of the CSV file you tested this with? I tested this now with only one email and it seems to work..

![image](https://user-images.githubusercontent.com/45761921/203821600-53824c68-959e-4487-b70e-4f8f114c3bca.png)

maximiliansoelch reviewed (changes requested):
You should also add the email in all the dialog texts and error messages:

> The following rows are invalid and must contain a value in either login, matriculation number **or email**:

evgeny1984 commented:
> > Currently, only works for > 1 email entries. When only 1 email is in CSV file, the following error is displayed: ![image](https://user-images.githubusercontent.com/10652693/203321701-92dccd25-b1af-4733-9fc5-c26e12c4642d.png)
> 
> @evgeny1984 could you post the content of the CSV file you tested this with? I tested this now with only one email and it seems to work..
> 
> ![image](https://user-images.githubusercontent.com/45761921/203821600-53824c68-959e-4487-b70e-4f8f114c3bca.png)

Actually, I just used the export file from Artemis and only pasted the email address there to test. Nothing special.

daniels98it commented:
> You should also add the email in all the dialog texts and error messages:
> 
> > The following rows are invalid and must contain a value in either login, matriculation number **or email**:

Thanks @

> You should also add the email in all the dialog texts and error messages:
> 
> > The following rows are invalid and must contain a value in either login, matriculation number **or email**:

Thanks @maximiliansoelch 🙂 
Have now been added 👍

daniels98it commented:
> Actually, I just used the export file from Artemis and only pasted the email address there to test. Nothing special.

@evgeny1984 Hmm... strange, I tested it now with the exported users CSV (with just one user), where I removed all columns except for the "email" one and it worked.. Did you keep the "email" header in the CSV file?

Could you please re-test it or provide on which Test Server and user to import this is failing?

evgeny1984 reviewed (changes requested):
> > Actually, I just used the export file from Artemis and only pasted the email address there to test. Nothing special.
> 
> @evgeny1984 Hmm... strange, I tested it now with the exported users CSV (with just one user), where I removed all columns except for the "email" one and it worked.. Did you keep the "email" header in the CSV file?
> 
> Could you please re-test it or provide on which Test Server and user to import this is failing?

The import with 1 user works now. Even if there is only 1 column with email, but also with other empty columns as well.
If I have no header in column and try to import, then the button is disabled, but I don't get any information what is wrong. There should be some feedback provided by the system to the user. 
![image](https://user-images.githubusercontent.com/10652693/204318885-2eb2b868-afd5-4f65-9855-ad511dc61301.png)
Here is the csv file:
[test-user-import.csv](https://github.com/ls1intum/Artemis/files/10105206/test-user-import.csv)

daniels98it commented:
> > > Actually, I just used the export file from Artemis and only pasted the email address there to test. Nothing special.
> > 
> > 
> > @evgeny1984 Hmm... strange, I tested it now with the exported users CSV (with just one user), where I removed all columns except for the "email" one and it worked.. Did you keep the "email" header in the CSV file?
> > Could you please re-test it or provide on which Test Server and user to import this is failing?
> 
> The import with 1 user works now. Even if there is only 1 column with email, but also with other empty columns as well. If I have no header in column and try to import, then the button is disabled, but I don't get any information what is wrong. There should be some feedback provided by the system to the user. ![image](https://user-images.githubusercontent.com/10652693/204318885-2eb2b868-afd5-4f65-9855-ad511dc61301.png) Here is the csv file: [test-user-import.csv](https://github.com/ls1intum/Artemis/files/10105206/test-user-import.csv)

The problem with `test-user-import.csv` is that it's missing the headers on the first row of the file, leading to no users being found.. I now added an additional check for this specific case, reporting about no found users (see screenshot). I'll additionally add a small snippet of an example CSV file which should help with its formatting.

Regarding the use of the exported CSV file, I think headers could be automatically added but I would suggest to integrate this change in a different PR.

![image](https://user-images.githubusercontent.com/45761921/205602867-ab669594-6e33-4e07-ae6d-2fb7b87bb7ef.png)

danielbode reviewed (dismissed):
Tested on ts1. Works as expected!

evgeny1984 reviewed (approved):
Last changes resolved the issue I requested to change. Thanks.

stefanwaldhauser reviewed (approved):
Nice changes and works as expected!

daniels98it commented:
> Tested on TS1. Import works 👍 I just noticed one small issue: If you import users by username and the user does not exist in the system, the table row turns red for the not imported users. When I import users by email and some do not exist, it still shows a green row after clicking the import button: ![image](https://user-images.githubusercontent.com/1368405/205663755-b93a65db-09cd-4643-a94c-44c4f7cbf903.png)
> 
> I think the row should be red, if the user email is not found

Missed that one 😅 
Thanks @TheZoker for noticing 👍 Has been fixed with [2b86566](https://github.com/ls1intum/Artemis/pull/5722/commits/2b865663fd56c88cbfcbfb165dd82d908824b305)

evgeny1984 reviewed (approved):
Tested on TS2, works well, also last integrated changes with not existing users.

daniels98it commented:
> > # [Codecov](https://codecov.io/gh/ls1intum/Artemis/pull/5722?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum) Report
> > Base: **84.60**% // Head: **79.91**% // Decreases project coverage by **`-4.68%`** ⚠️
> > > Coverage data is based on head [(`6c11cff`)](https://codecov.io/gh/ls1intum/Artemis/pull/5722?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum) compared to base [(`0bed62a`)](https://codecov.io/gh/ls1intum/Artemis/commit/0bed62afa61e4dd776c4df6285272f6151e898e5?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum).
> > > Patch coverage: 84.00% of modified lines in pull request are covered.
> > 
> > 
> > Additional details and impacted files
> > [☔ View full report at Codecov](https://codecov.io/gh/ls1intum/Artemis/pull/5722?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum). 📢 Do you have feedback about the report comment? [Let us know in this issue](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ls1intum).
> 
> This seems like a huge drop in coverage? @daniels98it Maybe you could write more tests?

This was probably an issue with Codecov as after fixing the last issue it no longer reports about such decrease in coverage..

krusche commented:
@daniels98it please resolve the merge conflicts

daniels98it commented:
The merge conflicts have been resolved. The functionality of this PR should have remained unchanged as the conflicts only related to the `CourseTestService.java` file.
